### PR TITLE
feat(gatsby): Reduce cost of sourcing after the initial

### PR DESCRIPTION
**Pull request by KyleAMathews**
We take some expensive steps after sourcing to warn plugin authors about possible errors and to
remove stale nodes from previous cached builds.

But these are no-ops on subsequent sources so we should skip them.

Calling getNodes() is especially expensive with LMDB (roughly 1s / 50k nodes on my machine). Deleting stale nodes is around 1s / 200k nodes. For one large site with ~700k nodes, making this change dropped sourcing time by ~15s.

### CODE REVIEW

**Comment by pragmaticpat** [comment]
[sc-41231]

test...

**Comment by pragmaticpat** [comment]
[sc-41231]

File: packages/gatsby/src/utils/source-nodes.ts [hunk 1]
```diff
@@ -104,14 +105,19 @@ export default async ({
...
 
-  deleteStaleNodes(state, nodes)
+    warnForPluginsWithoutNodes(state, nodes)
 
-  store.dispatch(actions.apiFinished({ apiName: `sourceNodes` }))
+    deleteStaleNodes(state, nodes)
+    initialSourcing = false
+  } else {
+    return
+  }
```
Comment (wardpeet) [inline packages/gatsby/src/utils/source-nodes.ts:122]: seems to be some unnecessary code?
```suggestion
  }
```

File: packages/gatsby/src/utils/source-nodes.ts [hunk 2]
```diff
@@ -85,6 +85,7 @@ function deleteStaleNodes(state: IGatsbyState, nodes: Array<Node>): void {
   }
 }
 
+let initialSourcing = true
```
Comment (wardpeet) [inline packages/gatsby/src/utils/source-nodes.ts:88]:
```suggestion
let isInitialSourcing = true
```

File: packages/gatsby/src/utils/source-nodes.ts [hunk 3]
```diff
@@ -104,14 +105,19 @@ export default async ({
...
+  // We only warn for plugins w/o nodes and delete stale nodes on the first sourcing.
+  if (initialSourcing) {
+    const state = store.getState()
+    const nodes = getNodes()
 
-  deleteStaleNodes(state, nodes)
+    warnForPluginsWithoutNodes(state, nodes)
 
-  store.dispatch(actions.apiFinished({ apiName: `sourceNodes` }))
+    deleteStaleNodes(state, nodes)
```
Comment (wardpeet) [inline packages/gatsby/src/utils/source-nodes.ts:118]: I guess we want to add a TODO to implement this later on as I'm sure we want to do this until we remove away from touchNodes.

```suggestion
    deleteStaleNodes(state, nodes)
```

Comment (KyleAMathews) [reply packages/gatsby/src/utils/source-nodes.ts:118]: implement what exactly?

Comment (KyleAMathews) [reply packages/gatsby/src/utils/source-nodes.ts:118]: deleting stale nodes on every call of sourceNodes?

Comment (wardpeet) [reply packages/gatsby/src/utils/source-nodes.ts:118]: I can't remember 😅 I also don't know what I was referring to. But shouldn't we keep dispatching the apifinished piece

**Review by wardpeet — DISMISSED** [review]

File: packages/gatsby/src/utils/source-nodes.ts (hunk 3 above)
Comment (KyleAMathews) [reply packages/gatsby/src/utils/source-nodes.ts:118]: oh I see what you mean now — the `store.dispatch(actions.apiFinished({ apiName: `sourceNodes` }))` line — hmm yeah didn't mean to remove that. It didn't seem to break anything either haha but I'll add it back

Comment (wardpeet) [reply packages/gatsby/src/utils/source-nodes.ts:118]: Great that you figured it out! 👍  We need that for Gatsby Cloud engines uploading

Comment (KyleAMathews) [reply packages/gatsby/src/utils/source-nodes.ts:118]: Ah sweet 👍 — stab the approve button when you have a chance and lets get this in!

**Review by wardpeet — APPROVED** [review]

**Pull request merged**